blog/2023-05-02-understanding-context-switching-and-its-impact-on-system-performance/index.md

```diff
@@ -0,0 +1,52 @@
+---
+slug: understanding-context-switching-and-its-impact-on-system-performance
+title: Understanding Context Switching and Its Impact on System Performance
+authors: satya
+tags: [processes, context-switch, performance, monitoring]
+keywords: [processes, context-switch, performance, monitoring]
+image: ./img/stacked-netdata.png
+---
+
+![stacked-netdata](./img/stacked-netdata.png)
+
+Context switching is the process of switching the CPU from one process, task or thread to another. In a multitasking operating system, such as Linux, the CPU has to switch between multiple processes or threads in order to keep the system running smoothly. This is necessary because the CPU can only execute one process or thread at a time. If there are many processes or threads running simultaneously, and very few CPU cores available to handle them, the system is forced to make more context switches to balance the CPU resources among them.
+
+Context switching is an essential function of any multitasking operating system, but it also comes at a cost. The whole process is computationally intensive, and the more context switches that occur, the slower the system becomes. This is because each context switch involves saving the current state of the CPU, loading the state of the new process or thread, and then resuming execution of the new process or thread. This takes time and consumes CPU resources, which can slow down the system.
+
+The impact of context switching on system performance can be significant, especially in systems with many processes or threads running simultaneously.
+
+<!-- truncate -->
+
+## Thrashing
+
+Thrashing is a phenomenon where the system spends more time switching between processes or threads than actually executing them. This can cause the system to become unresponsive and slow to the point of being unusable. Detecting thrashing can be challenging, but there are a few signs that can indicate its presence.
+
+One way to detect thrashing is by monitoring the system's performance metrics, such as CPU utilization, disk I/O, and memory usage. If these metrics are consistently high, while the system is slow or unresponsive, it may be an indication of thrashing.
+
+Another way to detect thrashing is by monitoring the number of context switches per second using tools like netdata. A sudden increase in context switches, especially when accompanied by a decrease in system performance, can be a sign of thrashing.
+
+In addition, monitoring the number of processes in the run queue (Runnable dimension in `system.processes_state chart`), the number of TASKLET and SCHED softirqs, as well as the time spent handling these softirqs, you can identify if there is a high level of task scheduling activity on the system.
```

Comment:
```suggestion
In addition, monitoring the number of processes in the run queue via the `runnable` dimension in the `system.processes_state chart`, the number of `TASKLET` and `SCHED` softirqs, as well as the time spent handling these softirqs, you can identify if there is a high level of task scheduling activity on the system.
```

Reply:
I very much like how we specifically mention the chart and dimension for the first parameter. Can we do do the same for the others too?